### PR TITLE
Prevent reentry to drainExpired loop

### PR DESCRIPTION
**Pull request by kriskowal**
r @jcorbin @raynos @rf

### CODE REVIEW

**Comment by jcorbin** [comment]
Tests?

File: time_heap.js [hunk 1]
```diff
@@ -157,8 +166,16 @@ TimeHeap.prototype.drainExpired = function drainExpired(now) {
         }
     }
     if (self.expired.length) {
+        // callExpiredTimeouts can re-enter drainExpired by calling popOutReq,
+        // which in turn calls update, and then back to drainExpired.
+        // The drainUntil shorts this recursive loop.
```
Comment (Raynos) [inline time_heap.js]: I feel like `nextTick()` was designed to avoid these kind of lulz.

File: time_heap.js [hunk 2]
```diff
@@ -150,6 +153,12 @@ TimeHeap.prototype.onTimeout = function onTimeout(now) {
 TimeHeap.prototype.drainExpired = function drainExpired(now) {
     var self = this;
 
+    if (now < self.drainUntil) {
```
Comment (Raynos) [inline time_heap.js]: Consider a boolean lock variable instead of a 100ms loop ttl.

File: time_heap.js [hunk 3]
```diff
@@ -157,8 +166,16 @@ TimeHeap.prototype.drainExpired = function drainExpired(now) {
         }
     }
     if (self.expired.length) {
+        // callExpiredTimeouts can re-enter drainExpired by calling popOutReq,
+        // which in turn calls update, and then back to drainExpired.
+        // The drainUntil shorts this recursive loop.
+        // The solution uses a timer instead of a flag to ensure that drain
+        // resumes even if an exception halts the loop prematurely without
```
Comment (Raynos) [inline time_heap.js]: LOL. this is why uncaught exception == you better restart the process.

**Comment by Raynos** [comment]
lgtm. Lets get this in.

**Pull request merged**